In [1]:
import pandas as pd

<h3>Collecting and loading the dataset</h3>

*  I have collected two datasets containing details about both phishing and legitimate websites.
*  Each dataset includes features of the URLs along with corresponding labels:
     *  A label of 0 indicates a legitimate website.
     *  A label of 1 indicates a phishing website.
* The data is based on HTML content and includes various features extracted from the URLs.
 URLs

In [2]:
legit = pd.read_csv("structured_data_legitimate.csv")

In [3]:
legit.head()

has_title  has_input  has_button  has_image  has_submit  has_link  \
0          1          1           0          0           1         0   
1          1          1           0          0           0         1   
2          1          1           1          0           0         1   
3          1          1           1          0           0         1   
4          1          1           1          0           0         1   

   has_password  has_email_input  has_hidden_element  has_audio  ...  \
0             0                0                   1          0  ...   
1             0                0                   0          0  ...   
2             1                0                   1          0  ...   
3             0                0                   0          0  ...   
4             0                1                   0          0  ...   

   has_text_input  number_of_meta  has_nav  has_object  has_picture  \
0               0               2        0           0            0   
1               1               7        0           0            0   
2               1              18        0           0            0   
3               0              26        1           0            1   
4               0              20        0           0            0   

   number_of_sources  number_of_span  number_of_table                   URL  \
0                  0               8                1     http://google.com   
1                  0               1                0    http://youtube.com   
2                  0               2                0   http://facebook.com   
3                 44              27                0  http://microsoft.com   
4                  2              31                0    http://netflix.com   

   label  
0      0  
1      0  
2      0  
3      0  
4      0  

[5 rows x 45 columns]

In [4]:
phish = pd.read_csv("structured_data_phishing.csv")

In [5]:
phish.head()

has_title  has_input  has_button  has_image  has_submit  has_link  \
0          0          0           0          0           0         0   
1          0          0           0          0           0         0   
2          0          0           0          0           0         0   
3          0          0           0          0           0         0   
4          0          0           0          0           0         0   

   has_password  has_email_input  has_hidden_element  has_audio  ...  \
0             0                0                   0          0  ...   
1             0                0                   0          0  ...   
2             0                0                   0          0  ...   
3             0                0                   0          0  ...   
4             0                0                   0          0  ...   

   has_text_input  number_of_meta  has_nav  has_object  has_picture  \
0               0               0        0           0            0   
1               0               0        0           0            0   
2               0               0        0           0            0   
3               0               0        0           0            0   
4               0               0        0           0            0   

   number_of_sources  number_of_span  number_of_table  \
0                  0               0                0   
1                  0               0                0   
2                  0               0                0   
3                  0               0                0   
4                  0               0                0   

                                                 URL  label  
0  http://www.eki-net.con-aescceccesaas.qfuhtb.to...      1  
1  http://www.eki-net.con-aescceccesaas.qsnbpy.to...      1  
2  http://www.eki-net.con-aescceccesaas.trea.top/...      1  
3  http://www.eki-net.con-aescceccesaas.ucwxyw.to...      1  
4  http://www.eki-net.con-aescceccesaas.vcuu.top/...      1  

[5 rows x 45 columns]

In [6]:
phish.shape

(10524, 45)

In [82]:
legit.shape

(16061, 50)

In [7]:
legit.columns

Index(['has_title', 'has_input', 'has_button', 'has_image', 'has_submit',
       'has_link', 'has_password', 'has_email_input', 'has_hidden_element',
       'has_audio', 'has_video', 'number_of_inputs', 'number_of_buttons',
       'number_of_images', 'number_of_option', 'number_of_list',
       'number_of_th', 'number_of_tr', 'number_of_href', 'number_of_paragraph',
       'number_of_script', 'length_of_title', 'has_h1', 'has_h2', 'has_h3',
       'length_of_text', 'number_of_clickable_button', 'number_of_a',
       'number_of_img', 'number_of_div', 'number_of_figure', 'has_footer',
       'has_form', 'has_text_area', 'has_iframe', 'has_text_input',
       'number_of_meta', 'has_nav', 'has_object', 'has_picture',
       'number_of_sources', 'number_of_span', 'number_of_table', 'URL',
       'label'],
      dtype='object')

<h4>Changing the order of the URL column</h4>

*   I am rearranging the order of the URL column to facilitate the addition of new features to the end
d

In [84]:
url_column = legit['URL']

In [9]:
legit.drop(columns=["URL"],inplace=True)

In [10]:
legit.insert(0,"URL",url_column)


In [11]:
legit

URL  has_title  has_input  has_button  has_image  \
0            http://google.com          1          1           0          0   
1           http://youtube.com          1          1           0          0   
2          http://facebook.com          1          1           1          0   
3         http://microsoft.com          1          1           1          0   
4           http://netflix.com          1          1           1          0   
...                        ...        ...        ...         ...        ...   
16056       http://emis.gov.eg          1          1           0          0   
16057     http://hilton.com.tr          1          1           1          0   
16058  http://medicines.org.uk          1          1           1          0   
16059        http://piwigo.org          1          0           1          0   
16060        http://koddos.com          1          0           0          0   

       has_submit  has_link  has_password  has_email_input  \
0               1         0             0                0   
1               0         1             0                0   
2               0         1             1                0   
3               0         1             0                0   
4               0         1             0                1   
...           ...       ...           ...              ...   
16056           0         1             0                0   
16057           0         1             0                0   
16058           0         1             0                0   
16059           0         1             0                0   
16060           0         0             0                0   

       has_hidden_element  ...  has_iframe  has_text_input  number_of_meta  \
0                       1  ...           0               0               2   
1                       0  ...           1               1               7   
2                       1  ...           0               1              18   
3                       0  ...           1               0              26   
4                       0  ...           0               0              20   
...                   ...  ...         ...             ...             ...   
16056                   1  ...           0               0               1   
16057                   1  ...           0               0               5   
16058                   0  ...           1               1               6   
16059                   0  ...           0               0               4   
16060                   0  ...           0               0               0   

       has_nav  has_object  has_picture  number_of_sources  number_of_span  \
0            0           0            0                  0               8   
1            0           0            0                  0               1   
2            0           0            0                  0               2   
3            1           0            1                 44              27   
4            0           0            0                  2              31   
...        ...         ...          ...                ...             ...   
16056        1           0            0                  0               4   
16057        1           0            1                 22               9   
16058        1           0            0                  0              27   
16059        1           0            0                  0              21   
16060        0           0            0                  0               0   

       number_of_table  label  
0                    1      0  
1                    0      0  
2                    0      0  
3                    0      0  
4                    0      0  
...                ...    ...  
16056                2      0  
16057                0      0  
16058                0      0  
16059                0      0  
16060                0      0  

[16061 rows x 45 columns]

In [20]:
phish_url = phish['URL']

In [21]:
phish.drop(columns=['URL'],inplace=True)

In [22]:
phish.insert(0,'URL',phish_url)

In [23]:
phish

URL  has_title  \
0      http://www.eki-net.con-aescceccesaas.qfuhtb.to...          0   
1      http://www.eki-net.con-aescceccesaas.qsnbpy.to...          0   
2      http://www.eki-net.con-aescceccesaas.trea.top/...          0   
3      http://www.eki-net.con-aescceccesaas.ucwxyw.to...          0   
4      http://www.eki-net.con-aescceccesaas.vcuu.top/...          0   
...                                                  ...        ...   
10519                              https://dloodles.app/          1   
10520                             https://artblooks.xyz/          1   
10521                           http://ohnizwl.htswyx.cn          1   
10522                https://poocoinect-tk.blogspot.com/          1   
10523                    https://osmosis.sezenradyo.com/          1   

       has_input  has_button  has_image  has_submit  has_link  has_password  \
0              0           0          0           0         0             0   
1              0           0          0           0         0             0   
2              0           0          0           0         0             0   
3              0           0          0           0         0             0   
4              0           0          0           0         0             0   
...          ...         ...        ...         ...       ...           ...   
10519          0           0          0           0         0             0   
10520          0           0          0           0         1             0   
10521          0           0          0           0         0             0   
10522          0           1          0           0         1             0   
10523          0           1          0           0         1             0   

       has_email_input  has_hidden_element  ...  has_iframe  has_text_input  \
0                    0                   0  ...           0               0   
1                    0                   0  ...           0               0   
2                    0                   0  ...           0               0   
3                    0                   0  ...           0               0   
4                    0                   0  ...           0               0   
...                ...                 ...  ...         ...             ...   
10519                0                   0  ...           0               0   
10520                0                   0  ...           0               0   
10521                0                   0  ...           0               0   
10522                0                   0  ...           0               0   
10523                0                   0  ...           0               0   

       number_of_meta  has_nav  has_object  has_picture  number_of_sources  \
0                   0        0           0            0                  0   
1                   0        0           0            0                  0   
2                   0        0           0            0                  0   
3                   0        0           0            0                  0   
4                   0        0           0            0                  0   
...               ...      ...         ...          ...                ...   
10519               0        0           0            0                  0   
10520              11        0           0            0                  0   
10521               1        0           0            0                  0   
10522               4        1           0            0                  0   
10523               3        1           0            0                  0   

       number_of_span  number_of_table  label  
0                   0                0      1  
1                   0                0      1  
2                   0                0      1  
3                   0                0      1  
4                   0                0      1  
...               ...              ...    ...  
10519               0        

In [24]:
phish_url.head()

0    http://www.eki-net.con-aescceccesaas.qfuhtb.to...
1    http://www.eki-net.con-aescceccesaas.qsnbpy.to...
2    http://www.eki-net.con-aescceccesaas.trea.top/...
3    http://www.eki-net.con-aescceccesaas.ucwxyw.to...
4    http://www.eki-net.con-aescceccesaas.vcuu.top/...
Name: URL, dtype: object

In [25]:
phish_url.tail(10)

10514                                  http://Shido.market
10515                                https://tmac.finance/
10516                                 https://web-hft.com/
10517                             https://tiny.one/pznd63c
10518    https://synslabs.firebaseapp.com/wallet/connec...
10519                                https://dloodles.app/
10520                               https://artblooks.xyz/
10521                             http://ohnizwl.htswyx.cn
10522                  https://poocoinect-tk.blogspot.com/
10523                      https://osmosis.sezenradyo.com/
Name: URL, dtype: object

<h4>Importing packages needed to extract features from the url's</h4>

In [ ]:
from urllib.parse import urlparse,urlencode
import ipaddress


In [30]:
from bs4 import BeautifulSoup
import requests
from requests.adapters import HTTPAdapter
from requests.packages.urllib3.util.retry import Retry
from urllib3.exceptions import InsecureRequestWarning
from urllib3 import disable_warnings
import urllib.request

In [ ]:
legit_len_lesser_url

<h3>Length of URL</h3>

*  Computes the length of the URL. Phishers can use long URL to hide the doubtful part in the address bar. In this project, if the length of the URL is    greater than or equal 54 characters then the URL classified as phishing otherwise legitimate.
*  If the length of URL >= 54 , the value assigned to this feature is 1 (phishing) or else 0 (legitimate).

In [35]:
def get_url_length(url):
    if len(url) >= 54:
        return 1
    else:
        return 0

In [36]:
phish["url_length"] = phish["URL"].apply(get_url_length)

In [37]:
phish

URL  has_title  \
0      http://www.eki-net.con-aescceccesaas.qfuhtb.to...          0   
1      http://www.eki-net.con-aescceccesaas.qsnbpy.to...          0   
2      http://www.eki-net.con-aescceccesaas.trea.top/...          0   
3      http://www.eki-net.con-aescceccesaas.ucwxyw.to...          0   
4      http://www.eki-net.con-aescceccesaas.vcuu.top/...          0   
...                                                  ...        ...   
10519                              https://dloodles.app/          1   
10520                             https://artblooks.xyz/          1   
10521                           http://ohnizwl.htswyx.cn          1   
10522                https://poocoinect-tk.blogspot.com/          1   
10523                    https://osmosis.sezenradyo.com/          1   

       has_input  has_button  has_image  has_submit  has_link  has_password  \
0              0           0          0           0         0             0   
1              0           0          0           0         0             0   
2              0           0          0           0         0             0   
3              0           0          0           0         0             0   
4              0           0          0           0         0             0   
...          ...         ...        ...         ...       ...           ...   
10519          0           0          0           0         0             0   
10520          0           0          0           0         1             0   
10521          0           0          0           0         0             0   
10522          0           1          0           0         1             0   
10523          0           1          0           0         1             0   

       has_email_input  has_hidden_element  ...  has_text_input  \
0                    0                   0  ...               0   
1                    0                   0  ...               0   
2                    0                   0  ...               0   
3                    0                   0  ...               0   
4                    0                   0  ...               0   
...                ...                 ...  ...             ...   
10519                0                   0  ...               0   
10520                0                   0  ...               0   
10521                0                   0  ...               0   
10522                0                   0  ...               0   
10523                0                   0  ...               0   

       number_of_meta  has_nav  has_object  has_picture  number_of_sources  \
0                   0        0           0            0                  0   
1                   0        0           0            0                  0   
2                   0        0           0            0                  0   
3                   0        0           0            0                  0   
4                   0        0           0            0                  0   
...               ...      ...         ...          ...                ...   
10519               0        0           0            0                  0   
10520              11        0           0            0                  0   
10521               1        0           0            0                  0   
10522               4        1           0            0                  0   
10523               3        1           0            0                  0   

       number_of_span  number_of_table  label  url_length  
0                   0                0      1           1  
1                   0                0      1           1  
2                   0                0      1           0  
3                   0                0      1           1  
4                   0                0      1           0  
...               ...              ...    ...         ...  
10519               0                1      1           0  
10520              14        

In [38]:
phish["url_length"].value_counts()

url_length
1    6414
0    4110
Name: count, dtype: int64

In [39]:
legit["url_length"] = legit["URL"].apply(get_url_length)

In [40]:
legit

URL  has_title  has_input  has_button  has_image  \
0            http://google.com          1          1           0          0   
1           http://youtube.com          1          1           0          0   
2          http://facebook.com          1          1           1          0   
3         http://microsoft.com          1          1           1          0   
4           http://netflix.com          1          1           1          0   
...                        ...        ...        ...         ...        ...   
16056       http://emis.gov.eg          1          1           0          0   
16057     http://hilton.com.tr          1          1           1          0   
16058  http://medicines.org.uk          1          1           1          0   
16059        http://piwigo.org          1          0           1          0   
16060        http://koddos.com          1          0           0          0   

       has_submit  has_link  has_password  has_email_input  \
0               1         0             0                0   
1               0         1             0                0   
2               0         1             1                0   
3               0         1             0                0   
4               0         1             0                1   
...           ...       ...           ...              ...   
16056           0         1             0                0   
16057           0         1             0                0   
16058           0         1             0                0   
16059           0         1             0                0   
16060           0         0             0                0   

       has_hidden_element  ...  has_text_input  number_of_meta  has_nav  \
0                       1  ...               0               2        0   
1                       0  ...               1               7        0   
2                       1  ...               1              18        0   
3                       0  ...               0              26        1   
4                       0  ...               0              20        0   
...                   ...  ...             ...             ...      ...   
16056                   1  ...               0               1        1   
16057                   1  ...               0               5        1   
16058                   0  ...               1               6        1   
16059                   0  ...               0               4        1   
16060                   0  ...               0               0        0   

       has_object  has_picture  number_of_sources  number_of_span  \
0               0            0                  0               8   
1               0            0                  0               1   
2               0            0                  0               2   
3               0            1                 44              27   
4               0            0                  2              31   
...           ...          ...                ...             ...   
16056           0            0                  0               4   
16057           0            1                 22               9   
16058           0            0                  0              27   
16059           0            0                  0              21   
16060           0            0                  0               0   

       number_of_table  label  url_length  
0                    1      0           0  
1                    0      0           0  
2                    0      0           0  
3                    0      0           0  
4                    0      0           0  
...                ...    ...         ...  
16056                2      0           0  
16057                0      0           0  
16058                0      0           0  
16059                0      0           0  
16060                0      0           0  

[16061 rows x 46 columns]

In [41]:
legit['url_length'].value_counts()

url_length
0    16060
1        1
Name: count, dtype: int64

 <h5>Data story</h5>

*  In the legitimate URLs, the majority (16060 instances) have a URL length less than a certain threshold, while only 1 instance has a URL               length greater than the threshold.
*  In the legitimate URLs, the majority (6414 instances) have a URL length greater than a certain threshold, while only (4110 instance) has a URL               length less than the thresholdd.

<h3>Prefix or Suffix "-" in Domain</h3>

*  Checking the presence of '-' in the domain part of URL. The dash symbol is rarely used in legitimate URLs. Phishers tend to add prefixes or suffixes    separated by (-) to the domain name so that users feel that they are dealing with a legitimate webpage.
*  If the URL has '-' symbol in the domain part of the URL, the value assigned to this feature is 1 (phishing) or else 0 (legitimate).

In [87]:
def has_hyphens(url):
    if '-' in urlparse(url).netloc: # we will be using the urlparse from urllib.parse for checking the presence of '-' in the url
        return 1
    else:
        return 0

In [43]:
legit["has_hyphens"] = legit["URL"].apply(has_hyphens)

In [44]:
legit

URL  has_title  has_input  has_button  has_image  \
0            http://google.com          1          1           0          0   
1           http://youtube.com          1          1           0          0   
2          http://facebook.com          1          1           1          0   
3         http://microsoft.com          1          1           1          0   
4           http://netflix.com          1          1           1          0   
...                        ...        ...        ...         ...        ...   
16056       http://emis.gov.eg          1          1           0          0   
16057     http://hilton.com.tr          1          1           1          0   
16058  http://medicines.org.uk          1          1           1          0   
16059        http://piwigo.org          1          0           1          0   
16060        http://koddos.com          1          0           0          0   

       has_submit  has_link  has_password  has_email_input  \
0               1         0             0                0   
1               0         1             0                0   
2               0         1             1                0   
3               0         1             0                0   
4               0         1             0                1   
...           ...       ...           ...              ...   
16056           0         1             0                0   
16057           0         1             0                0   
16058           0         1             0                0   
16059           0         1             0                0   
16060           0         0             0                0   

       has_hidden_element  ...  number_of_meta  has_nav  has_object  \
0                       1  ...               2        0           0   
1                       0  ...               7        0           0   
2                       1  ...              18        0           0   
3                       0  ...              26        1           0   
4                       0  ...              20        0           0   
...                   ...  ...             ...      ...         ...   
16056                   1  ...               1        1           0   
16057                   1  ...               5        1           0   
16058                   0  ...               6        1           0   
16059                   0  ...               4        1           0   
16060                   0  ...               0        0           0   

       has_picture  number_of_sources  number_of_span  number_of_table  label  \
0                0                  0               8                1      0   
1                0                  0               1                0      0   
2                0                  0               2                0      0   
3                1                 44              27                0      0   
4                0                  2              31                0      0   
...            ...                ...             ...              ...    ...   
16056            0                  0               4                2      0   
16057            1                 22               9                0      0   
16058            0                  0              27                0      0   
16059            0                  0              21                0      0   
16060            0                  0               0                0      0   

       url_length  has_hyphens  
0               0            0  
1               0            0  
2               0            0  
3               0            0  
4               0            0  
...           ...          ...  
16056           0            0  
16057           0            0  
16058           0            0  
16059           0            0  
16060           0            0  

[16061 rows x 47 columns]

In [45]:
legit["has_hyphens"].value_counts()

has_hyphens
0    15466
1      595
Name: count, dtype: int64

In [46]:
phish["has_hyphens"] = phish["URL"].apply(has_hyphens)

In [47]:
phish

URL  has_title  \
0      http://www.eki-net.con-aescceccesaas.qfuhtb.to...          0   
1      http://www.eki-net.con-aescceccesaas.qsnbpy.to...          0   
2      http://www.eki-net.con-aescceccesaas.trea.top/...          0   
3      http://www.eki-net.con-aescceccesaas.ucwxyw.to...          0   
4      http://www.eki-net.con-aescceccesaas.vcuu.top/...          0   
...                                                  ...        ...   
10519                              https://dloodles.app/          1   
10520                             https://artblooks.xyz/          1   
10521                           http://ohnizwl.htswyx.cn          1   
10522                https://poocoinect-tk.blogspot.com/          1   
10523                    https://osmosis.sezenradyo.com/          1   

       has_input  has_button  has_image  has_submit  has_link  has_password  \
0              0           0          0           0         0             0   
1              0           0          0           0         0             0   
2              0           0          0           0         0             0   
3              0           0          0           0         0             0   
4              0           0          0           0         0             0   
...          ...         ...        ...         ...       ...           ...   
10519          0           0          0           0         0             0   
10520          0           0          0           0         1             0   
10521          0           0          0           0         0             0   
10522          0           1          0           0         1             0   
10523          0           1          0           0         1             0   

       has_email_input  has_hidden_element  ...  number_of_meta  has_nav  \
0                    0                   0  ...               0        0   
1                    0                   0  ...               0        0   
2                    0                   0  ...               0        0   
3                    0                   0  ...               0        0   
4                    0                   0  ...               0        0   
...                ...                 ...  ...             ...      ...   
10519                0                   0  ...               0        0   
10520                0                   0  ...              11        0   
10521                0                   0  ...               1        0   
10522                0                   0  ...               4        1   
10523                0                   0  ...               3        1   

       has_object  has_picture  number_of_sources  number_of_span  \
0               0            0                  0               0   
1               0            0                  0               0   
2               0            0                  0               0   
3               0            0                  0               0   
4               0            0                  0               0   
...           ...          ...                ...             ...   
10519           0            0                  0               0   
10520           0            0                  0              14   
10521           0            0                  0               0   
10522           0            0                  0              51   
10523           0            0                  0               0   

       number_of_table  label  url_length  has_hyphens  
0                    0      1           1            1  
1                    0      1           1            1  
2                    0      1           0            1  
3                    0      1           1            1  
4                    0      1           0            1  
...                ...    ...         ...          ...  
10519                1      1           0            0  
10520                0      1           0            

In [48]:
phish['has_hyphens'].value_counts()

has_hyphens
0    5794
1    4730
Name: count, dtype: int64

<h5>Data story</h5> 
   
   * In the phishing URLs, there are 5794 rows without '-' and 4730 rows with '-' hyphens.
   * In the legitimate URLs, there are 15466 rows without hyphens '-' and 595 rows with hyphens '-'
   * Therefore, hyphen checks will be a crucial feature for training the machine learning model in this project

In [49]:
import ipaddress

def having_ip(url):
    try:
        parsed_url = urlparse(url)
        ipaddress.ip_address(parsed_url.netloc)
        return 1
    except ValueError:
        return 0


In [50]:
phish["having_ip"] = phish["URL"].apply(having_ip)

In [51]:
phish["having_ip"].value_counts()

having_ip
0    10520
1        4
Name: count, dtype: int64

In [52]:
legit["having_ip"] = legit["URL"].apply(having_ip)

In [53]:
legit['having_ip'].value_counts()

having_ip
0    16061
Name: count, dtype: int64

In [ ]:
"""legit.drop('having_ip',inplace=True,axis=1)"""

In [ ]:
"""legit"""

In [ ]:
"""phish.drop('having_ip',inplace=True,axis=1)"""

In [ ]:
"""phish"""

In [56]:
def having_atsign(url):
    if '@' in urlparse(url).netloc:
        return 1
    else:
        return 0

In [57]:
phish["atsign"] = phish['URL'].apply(having_atsign)

In [58]:
phish["atsign"].value_counts()

atsign
0    10524
Name: count, dtype: int64

In [59]:
legit["atsign"] = legit['URL'].apply(having_atsign)

In [60]:
legit['atsign'].value_counts()

atsign
0    16061
Name: count, dtype: int64

In [ ]:
"""legit.drop("atsign",inplace=True,axis=1)"""

In [ ]:
"""phish.drop("atsign",inplace=True,axis=1)"""

In [ ]:
"""legit.head()"""

In [ ]:
"""def redirection(url):
    last_slash = url.rfind("//")
    if last_slash > 7:
        return 1
    else:
        return 0"""

In [ ]:
"""phish["redirection"] = phish['URL'].apply(redirection)"""

In [ ]:
"""phish["redirection"].value_counts()"""

In [ ]:
"""phish.drop("redirection",inplace=True,axis=1)"""

In [61]:
def has_http_in_domain(url):
    parsed_url = urlparse(url)
    domain = parsed_url.netloc.split(':')[0]
    if 'http' in domain:
        return 1
    else:
        return 0

In [62]:
phish['http_in_domain'] = phish['URL'].apply(has_http_in_domain)


In [63]:
phish['http_in_domain'].value_counts()

http_in_domain
0    10524
Name: count, dtype: int64

In [64]:
phish

URL  has_title  \
0      http://www.eki-net.con-aescceccesaas.qfuhtb.to...          0   
1      http://www.eki-net.con-aescceccesaas.qsnbpy.to...          0   
2      http://www.eki-net.con-aescceccesaas.trea.top/...          0   
3      http://www.eki-net.con-aescceccesaas.ucwxyw.to...          0   
4      http://www.eki-net.con-aescceccesaas.vcuu.top/...          0   
...                                                  ...        ...   
10519                              https://dloodles.app/          1   
10520                             https://artblooks.xyz/          1   
10521                           http://ohnizwl.htswyx.cn          1   
10522                https://poocoinect-tk.blogspot.com/          1   
10523                    https://osmosis.sezenradyo.com/          1   

       has_input  has_button  has_image  has_submit  has_link  has_password  \
0              0           0          0           0         0             0   
1              0           0          0           0         0             0   
2              0           0          0           0         0             0   
3              0           0          0           0         0             0   
4              0           0          0           0         0             0   
...          ...         ...        ...         ...       ...           ...   
10519          0           0          0           0         0             0   
10520          0           0          0           0         1             0   
10521          0           0          0           0         0             0   
10522          0           1          0           0         1             0   
10523          0           1          0           0         1             0   

       has_email_input  has_hidden_element  ...  has_picture  \
0                    0                   0  ...            0   
1                    0                   0  ...            0   
2                    0                   0  ...            0   
3                    0                   0  ...            0   
4                    0                   0  ...            0   
...                ...                 ...  ...          ...   
10519                0                   0  ...            0   
10520                0                   0  ...            0   
10521                0                   0  ...            0   
10522                0                   0  ...            0   
10523                0                   0  ...            0   

       number_of_sources  number_of_span  number_of_table  label  url_length  \
0                      0               0                0      1           1   
1                      0               0                0      1           1   
2                      0               0                0      1           0   
3                      0               0                0      1           1   
4                      0               0                0      1           0   
...                  ...             ...              ...    ...         ...   
10519                  0               0                1      1           0   
10520                  0              14                0      1           0   
10521                  0               0                0      1           0   
10522                  0              51                0      1           0   
10523                  0               0                0      1           0   

       has_hyphens  having_ip  atsign  http_in_domain  
0                1          0       0               0  
1                1          0       0               0  
2                1          0       0               0  
3                1          0       0               0  
4                1          0       0               0  
...            ...        ...     ...             ...  
10519            0          0       0               0  
10520            0          0       0               0  
10521            

In [65]:
phish.drop('http_in_domain',inplace=True,axis=1)

In [66]:
def extract_domain(url):
    parsed_url = urlparse(url)
    return parsed_url.netloc

In [67]:
legit['domain'] = legit["URL"].apply(extract_domain)

In [68]:
legit

URL  has_title  has_input  has_button  has_image  \
0            http://google.com          1          1           0          0   
1           http://youtube.com          1          1           0          0   
2          http://facebook.com          1          1           1          0   
3         http://microsoft.com          1          1           1          0   
4           http://netflix.com          1          1           1          0   
...                        ...        ...        ...         ...        ...   
16056       http://emis.gov.eg          1          1           0          0   
16057     http://hilton.com.tr          1          1           1          0   
16058  http://medicines.org.uk          1          1           1          0   
16059        http://piwigo.org          1          0           1          0   
16060        http://koddos.com          1          0           0          0   

       has_submit  has_link  has_password  has_email_input  \
0               1         0             0                0   
1               0         1             0                0   
2               0         1             1                0   
3               0         1             0                0   
4               0         1             0                1   
...           ...       ...           ...              ...   
16056           0         1             0                0   
16057           0         1             0                0   
16058           0         1             0                0   
16059           0         1             0                0   
16060           0         0             0                0   

       has_hidden_element  ...  has_picture  number_of_sources  \
0                       1  ...            0                  0   
1                       0  ...            0                  0   
2                       1  ...            0                  0   
3                       0  ...            1                 44   
4                       0  ...            0                  2   
...                   ...  ...          ...                ...   
16056                   1  ...            0                  0   
16057                   1  ...            1                 22   
16058                   0  ...            0                  0   
16059                   0  ...            0                  0   
16060                   0  ...            0                  0   

       number_of_span  number_of_table  label  url_length  has_hyphens  \
0                   8                1      0           0            0   
1                   1                0      0           0            0   
2                   2                0      0           0            0   
3                  27                0      0           0            0   
4                  31                0      0           0            0   
...               ...              ...    ...         ...          ...   
16056               4                2      0           0            0   
16057               9                0      0           0            0   
16058              27                0      0           0            0   
16059              21                0      0           0            0   
16060               0                0      0           0            0   

       having_ip  atsign            domain  
0              0       0        google.com  
1              0       0       youtube.com  
2              0       0      facebook.com  
3              0       0     microsoft.com  
4              0       0       netflix.com  
...          ...     ...               ...  
16056          0       0       emis.gov.eg  
16057          0       0     hilton.com.tr  
16058          0       0  medicines.org.uk  
16059          0       0        piwigo.org  
16060          0       0        koddos.com  

[16061 rows x 50 columns]

In [69]:
phish['domain'] = phish["URL"].apply(extract_domain)

In [86]:
phish

URL  has_title  \
0      http://www.eki-net.con-aescceccesaas.qfuhtb.to...          0   
1      http://www.eki-net.con-aescceccesaas.qsnbpy.to...          0   
2      http://www.eki-net.con-aescceccesaas.trea.top/...          0   
3      http://www.eki-net.con-aescceccesaas.ucwxyw.to...          0   
4      http://www.eki-net.con-aescceccesaas.vcuu.top/...          0   
...                                                  ...        ...   
10519                              https://dloodles.app/          1   
10520                             https://artblooks.xyz/          1   
10521                           http://ohnizwl.htswyx.cn          1   
10522                https://poocoinect-tk.blogspot.com/          1   
10523                    https://osmosis.sezenradyo.com/          1   

       has_input  has_button  has_image  has_submit  has_link  has_password  \
0              0           0          0           0         0             0   
1              0           0          0           0         0             0   
2              0           0          0           0         0             0   
3              0           0          0           0         0             0   
4              0           0          0           0         0             0   
...          ...         ...        ...         ...       ...           ...   
10519          0           0          0           0         0             0   
10520          0           0          0           0         1             0   
10521          0           0          0           0         0             0   
10522          0           1          0           0         1             0   
10523          0           1          0           0         1             0   

       has_email_input  has_hidden_element  ...  has_picture  \
0                    0                   0  ...            0   
1                    0                   0  ...            0   
2                    0                   0  ...            0   
3                    0                   0  ...            0   
4                    0                   0  ...            0   
...                ...                 ...  ...          ...   
10519                0                   0  ...            0   
10520                0                   0  ...            0   
10521                0                   0  ...            0   
10522                0                   0  ...            0   
10523                0                   0  ...            0   

       number_of_sources  number_of_span  number_of_table  label  url_length  \
0                      0               0                0      1           1   
1                      0               0                0      1           1   
2                      0               0                0      1           0   
3                      0               0                0      1           1   
4                      0               0                0      1           0   
...                  ...             ...              ...    ...         ...   
10519                  0               0                1      1           0   
10520                  0              14                0      1           0   
10521                  0               0                0      1           0   
10522                  0              51                0      1           0   
10523                  0               0                0      1           0   

       has_hyphens  having_ip  atsign  \
0                1          0       0   
1                1          0       0   
2                1          0       0   
3                1          0       0   
4                1          0       0   
...            ...        ...     ...   
10519            0          0       0   
10520            0          0       0   
10521            0          0       0   
10522            1          0       0   
10523            0          0       0   

                            

In [88]:
import asyncio
import aiohttp
import pandas as pd

async def fetch_website_traffic(url):
    async with aiohttp.ClientSession() as session:
        try:
            async with session.get(f"https://awis.api.alexa.com/api?Action=TrafficHistory&Range=31&ResponseGroup=History&Url={url}&Output=json") as response:
                data = await response.json()
                return url, data['TrafficHistory']
        except Exception as e:
            print(f"Error fetching traffic for {url}: {e}")
            return url, None

async def fetch_website_traffic_for_batch(urls):
    tasks = []
    async with aiohttp.ClientSession() as session:
        for url in urls:
            tasks.append(fetch_website_traffic(url))
        results = await asyncio.gather(*tasks)
        return results

async def main():
    phish_urls = phish['URL'].tolist()  # Extract URLs from DataFrame
    legit_urls = legit['URL'].tolist()  # Extract URLs from DataFrame

    # Define batch size for each DataFrame
    phish_batch_size = 5
    legit_batch_size = 5

    # Fetch website traffic for phishing URLs
    for i in range(0, len(phish_urls), phish_batch_size):
        batch_urls = phish_urls[i:i+phish_batch_size]
        results = await fetch_website_traffic_for_batch(batch_urls)
        for url, traffic in results:
            if traffic is not None:
                print(f"Traffic for phishing URL {url}: {traffic}")
            else:
                print(f"Failed to fetch traffic for phishing URL {url}")

    # Fetch website traffic for legitimate URLs
    for i in range(0, len(legit_urls), legit_batch_size):
        batch_urls = legit_urls[i:i+legit_batch_size]
        results = await fetch_website_traffic_for_batch(batch_urls)
        for url, traffic in results:
            if traffic is not None:
                print(f"Traffic for legitimate URL {url}: {traffic}")
            else:
                print(f"Failed to fetch traffic for legitimate URL {url}")

In [89]:
import asyncio
import aiohttp
import pandas as pd

async def fetch_alexa_rank(url):
    async with aiohttp.ClientSession() as session:
        try:
            async with session.get(f"https://awis.api.alexa.com/api?Action=UrlInfo&Url={url}&Output=json") as response:
                data = await response.json()
                return url, data['UrlInfo']['Rank']
        except Exception as e:
            print(f"Error fetching Alexa rank for {url}: {e}")
            return url, None

async def fetch_alexa_rank_for_batch(urls):
    tasks = []
    async with aiohttp.ClientSession() as session:
        for url in urls:
            tasks.append(fetch_alexa_rank(url))
        results = await asyncio.gather(*tasks)
        return results

async def main():
    phish_urls = phish['URL'].tolist()  # Extract URLs from DataFrame
    legit_urls = legit['URL'].tolist()  # Extract URLs from DataFrame

    # Define batch size for each DataFrame
    phish_batch_size = 5
    legit_batch_size = 5

    # Fetch Alexa rank for phishing URLs
    phish_rank = []
    for i in range(0, len(phish_urls), phish_batch_size):
        batch_urls = phish_urls[i:i+phish_batch_size]
        results = await fetch_alexa_rank_for_batch(batch_urls)
        phish_rank.extend(results)

    # Fetch Alexa rank for legitimate URLs
    legit_rank = []
    for i in range(0, len(legit_urls), legit_batch_size):
        batch_urls = legit_urls[i:i+legit_batch_size]
        results = await fetch_alexa_rank_for_batch(batch_urls)
        legit_rank.extend(results)

    # Add Alexa rank data to DataFrame
    phish_rank_df = pd.DataFrame(phish_rank, columns=['URL', 'Rank'])
    phish = pd.merge(phish, phish_rank_df, on='URL', how='left')
    phish['Alexa_Rank'] = phish['Rank'].apply(lambda x: 1 if x is not None and x < 100000 else 0)

    legit_rank_df = pd.DataFrame(legit_rank, columns=['URL', 'Rank'])
    legit = pd.merge(legit, legit_rank_df, on='URL', how='left')
    legit['Alexa_Rank'] = legit['Rank'].apply(lambda x: 1 if x is not None and x < 100000 else 0)

    # Save DataFrame with Alexa rank data to CSV files
    phish.to_csv('structured_data_phishing_with_rank.csv', index=False)
    legit.to_csv('structured_data_legitimate_with_rank.csv', index=False)

In [91]:
phish

URL  has_title  \
0      http://www.eki-net.con-aescceccesaas.qfuhtb.to...          0   
1      http://www.eki-net.con-aescceccesaas.qsnbpy.to...          0   
2      http://www.eki-net.con-aescceccesaas.trea.top/...          0   
3      http://www.eki-net.con-aescceccesaas.ucwxyw.to...          0   
4      http://www.eki-net.con-aescceccesaas.vcuu.top/...          0   
...                                                  ...        ...   
10519                              https://dloodles.app/          1   
10520                             https://artblooks.xyz/          1   
10521                           http://ohnizwl.htswyx.cn          1   
10522                https://poocoinect-tk.blogspot.com/          1   
10523                    https://osmosis.sezenradyo.com/          1   

       has_input  has_button  has_image  has_submit  has_link  has_password  \
0              0           0          0           0         0             0   
1              0           0          0           0         0             0   
2              0           0          0           0         0             0   
3              0           0          0           0         0             0   
4              0           0          0           0         0             0   
...          ...         ...        ...         ...       ...           ...   
10519          0           0          0           0         0             0   
10520          0           0          0           0         1             0   
10521          0           0          0           0         0             0   
10522          0           1          0           0         1             0   
10523          0           1          0           0         1             0   

       has_email_input  has_hidden_element  ...  has_picture  \
0                    0                   0  ...            0   
1                    0                   0  ...            0   
2                    0                   0  ...            0   
3                    0                   0  ...            0   
4                    0                   0  ...            0   
...                ...                 ...  ...          ...   
10519                0                   0  ...            0   
10520                0                   0  ...            0   
10521                0                   0  ...            0   
10522                0                   0  ...            0   
10523                0                   0  ...            0   

       number_of_sources  number_of_span  number_of_table  label  url_length  \
0                      0               0                0      1           1   
1                      0               0                0      1           1   
2                      0               0                0      1           0   
3                      0               0                0      1           1   
4                      0               0                0      1           0   
...                  ...             ...              ...    ...         ...   
10519                  0               0                1      1           0   
10520                  0              14                0      1           0   
10521                  0               0                0      1           0   
10522                  0              51                0      1           0   
10523                  0               0                0      1           0   

       has_hyphens  having_ip  atsign  \
0                1          0       0   
1                1          0       0   
2                1          0       0   
3                1          0       0   
4                1          0       0   
...            ...        ...     ...   
10519            0          0       0   
10520            0          0       0   
10521            0          0       0   
10522            1          0       0   
10523            0          0       0   

                            

In [92]:
df = pd.read_csv("structured_data_phishing_with_rank.csv")

FileNotFoundError: [Errno 2] No such file or directory: 'structured_data_phishing_with_rank.csv'

In [93]:
import os
print(os.getcwd()) 

C:\Users\DELL
